rero_ils/modules/patrons/schemas/json.py

```diff
@@ -0,0 +1,136 @@
+    """Marshmallow schema for the `Template` metadata."""
+
+    pid = SanitizedUnicode()
+    schema = GenFunction(
+        load_only=True,
+        attribute='$schema',
+        data_key='$schema',
+        deserialize=schema_from_template
+    )
+    source = SanitizedUnicode()
+    local_codes = fields.List(SanitizedUnicode())
+    user_id = fields.Integer()
+    second_address = fields.Nested(PatronAddressSchema())
+    patron = fields.Dict()  # TODO : stricter implementation
+    libraries = fields.Nested(RefSchema, many=True)
+    roles = fields.List(SanitizedUnicode(validate=OneOf(UserRole.ALL_ROLES)))
+    notes = fields.Nested(PatronNoteSchema, many=True)
+
+    @pre_load
+    def remove_user_data(self, data, many, **kwargs):
+        """Removed data concerning User not Patron.
+
+        :param data: the data received from request.
+        :param many: is the `data` represent an array of schema object.
+        :param kwargs: any additional named arguments.
+        :return Data cleared from user profile information.
+        """
+        data = data if many else [data]
+        profile_fields = set(User.profile_fields + ['username', 'email'])
+        for record in data:
+            for field in profile_fields:
+                record.pop(field, None)
+        return data if many else data[0]
+
+    @validates('roles')
+    @http_applicable_method('POST')
+    def validate_role(self, data, **kwargs):
+        """Validate `roles` attribute through API request.
+
+        The `roles` attribute must be controlled to restrict some role
+        attribution/modification depending on the current logged user.
+
+        :param data: the `roles` attribute value.
+        :param kwargs: any additional named arguments.
+        :raises ValidationError: if error has detected on `roles` attribute
+        """
+        validate_role_changes(current_user, set(data))
+
+    @validates_schema
+    @http_applicable_method('PUT')
+    def validate_roles_changes(self, data, **kwargs):
+        """Validate `roles` changes through REST API request.
+
+        Updates on `roles` attribute is subject to restriction depending
+        on current connected user. Determine if `roles` field changes and if
+        these changes are allowed or not.
+
+        :param data: the json data to validate.
+        :param kwargs: additional named arguments.
+        :raises abort: if corresponding record is not found.
+        :raises ValidationError: if error has detected on role field
+        """
+        # Load DB record
+        db_record = Patron.get_record_by_pid(data.get('pid'))
+        if not db_record:
+            abort(404)
+
+        # Check if `roles` of the patron changed. If not, we can stop
+        # the validation process.
+        original_roles = set(db_record.get('roles', []))
+        data_roles = set(data.get('roles', []))
+        role_changes = original_roles.symmetric_difference(data_roles)
+        if not role_changes:
```

Comment:
Detecting if the roles have been changed seems as complex as `validate_role_changes`. Is it really required?